client.go

```diff
@@ -0,0 +1,449 @@
+	resp, err := cs.send(&reflectionv1.ServerReflectionRequest{
+		MessageRequest: &reflectionv1.ServerReflectionRequest_AllExtensionNumbersOfType{
+			AllExtensionNumbersOfType: string(messageName),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	respExtNumbers := resp.GetAllExtensionNumbersResponse()
+	if respExtNumbers == nil {
+		return nil, errWrongResponseType(resp, "all_extension_numbers")
+	}
+	extNumbers := make([]protoreflect.FieldNumber, len(respExtNumbers.ExtensionNumber))
+	for i, num := range respExtNumbers.ExtensionNumber {
+		extNumbers[i] = protoreflect.FieldNumber(num)
+	}
+	return extNumbers, nil
+}
+
+func (cs *ClientStream) getDescriptors(operation string, req *reflectionv1.ServerReflectionRequest) ([]*descriptorpb.FileDescriptorProto, error) {
+	resp, err := cs.send(req)
+	if err != nil {
+		return nil, err
+	}
+	respDescriptors := resp.GetFileDescriptorResponse()
+	if respDescriptors == nil {
+		return nil, errWrongResponseType(resp, operation)
+	}
+	descriptors := make([]*descriptorpb.FileDescriptorProto, len(respDescriptors.FileDescriptorProto))
+	for i, data := range respDescriptors.FileDescriptorProto {
+		fileDescriptor := &descriptorpb.FileDescriptorProto{}
+		if err := proto.Unmarshal(data, fileDescriptor); err != nil {
+			return nil, fmt.Errorf("reply to %s contained invalid descriptor proto: %w", operation, err)
+		}
+		descriptors[i] = fileDescriptor
+	}
+	return descriptors, nil
+}
+
+func (cs *ClientStream) send(req *reflectionv1.ServerReflectionRequest) (*reflectionv1.ServerReflectionResponse, error) {
+	req.Host = cs.host
+	// Sending on a bidi stream is usually thread-safe. But the replies are in the same order
+	// as the requests. So to prevent concurrent use from interleaving replies (which would
+	// require much more logic here to properly correlate replies with requests), we send and
+	// receive while holding the mutex. This means that this API does not support pipelining
+	// reflection requests (which could in theory reduce latency, but only when the client
+	// knows all of their requests up-front, which is rarely the case since subsequent calls
+	// often depend on the data in prior responses.
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+	for {
+		stream := cs.getStreamLocked()
+		if err := stream.Send(req); err != nil {
+			if errors.Is(err, io.EOF) {
+				// need to call Receive to get actual error code
+				_, recvErr := stream.Receive()
+				if recvErr != nil {
+					err = recvErr
+				}
+			}
+			if cs.shouldRetryLocked(err) {
+				continue
+			}
+			return nil, &streamError{err: err}
+		}
+		resp, err := stream.Receive()
+		if err != nil {
+			if cs.shouldRetryLocked(err) {
+				continue
+			}
+			return nil, &streamError{err: err}
+		}
+		if errResp := resp.GetErrorResponse(); errResp != nil {
+			return nil, connect.NewError(connect.Code(errResp.ErrorCode), errors.New(errResp.ErrorMessage))
```

Comment:
Would `connect.NewWireError` be more appropriate here?

Reply:
Doh, yes! Thanks for catching that.